Ella/fail on extra fields

erohmensing opened the pull request:
## TODO
* Create follow up issues for community program to add extra fields to spec and turn off the override flag

## What
* close #21258 
* Add behavior to CAT which, during schema validation, fails if we receive undeclared columns in records. This is configurable via the `fail_on_extra_columns` input parameter, which defaults to `true`.

## How
During the basic read test, during schema validation, artificially inject/overwrite `additionalProperties` on the top level of the schema to be set to `false`. Records that come in with extra properties will then fail the test. Don't do this if `fail_on_extra_columns` is set to True. 

In order to avoid this breaking CAT runs for all of our GA/beta connectors and making it hard for teams to deploy updates to those connectors, I ran this test locally on all connectors. Beta and GA connectors that failed specifically due to the new `Additional properties are not allowed` validation were updated so that  `fail_on_extra_columns: True` is set on their basic read tests.


## Recommended reading order
1. `connector-acceptance-tests/test_core.py`: introduction of new test logic 
2. `connector-acceptance-tests/unit_tests/test_core.py`: unit tests for new logic
2. `acceptance-test-config.yml` files: connectors which have been updated to bypass this logic for now until their specs are updated

## 🚨 User Impact 🚨
No end user impact. Connector developers might find that a connector starts failing CAT because of this test. If so, they can add a bypass or fix the behavior (we attempted to add as many bypasses as possible for faiiling connectors to avoid this sudden failing).

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

alafanechere reviewed (approved):
LGTM but please consider my comment about nested properties. 
Please bump the version in the dockerfile + update changelog if you want to publish a new version of CAT.
Thank you for this addition, I'm sure it'll help us detect API changes faster.

erohmensing commented on airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/utils/asserts.py:
```diff
@@ -44,22 +46,36 @@ def check(self, instance, format):
+    if it is set, or will create the property and set it to False if it does not exist.
+    """
+    enforced_schema = copy.deepcopy(json_schema)
+    dpath.util.new(enforced_schema, "additionalProperties", False)
```
I don't believe so - we are only concerned about extra fields with regards to their impact on column selection... as far as I know. We could certainly be stricter about nested properties in the future if we wanted to, but for this case, I wouldn't want a connector failing the nested one to stop it from being added to the column selection flag on cloud. 

But I'll check with them to understand if nested properties impact column selection - I'm not actually familar with the UX

alafanechere replied:
From last all hands demo I think they're not yet handling selection of nested properties but might want to do it soon.

erohmensing replied:
Yep - From @mfsiega-airbyte: 
> Column selection so far only looks at top level fields. If there's an unexpected sub property, we'll continue passing it along. We might want to iterate on this in the future, but for now column selection is all top level fields.

This to me means that we can run the tests locally for the next steps without worrying about nested fields - but it also seems like a good idea for future proofing. I could add a separate flag for failing on extra properties in nested fields?

evantahler replied:
I don't know if column-selection will /ever/ support nested property selection... and I'll bet if we do, it's a long way off.  I don't think you need to add any code for it today- maybe just a comment noting that this is only checking top-level schema

erohmensing commented on airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:
```diff
@@ -813,8 +819,8 @@ def test_read(
 
-        if should_validate_schema:
-            self._validate_schema(records=records, configured_catalog=configured_catalog)
+        if should_validate_schema:  # TODO Conditional logic
```
Yep - there was weird potential conditional logic here where I'd want to run the extra fields check even if `should_validate_schema` was false... without failing on other schema validation errors which would be weird. 

As far as I can see, only one connector makes use of `should_validate_schema: false`. For that reason I'd be in support of removing it and making it mandatory. But that also sort of goes against our idea that connectors can opt out of anything if they want to 🤔 but I think there have to be limitations to that

alafanechere replied:
> But that also sort of goes against our idea that connectors can opt out of anything if they want to 🤔

I'm not sure I'm very comfortable with this idea. I'd love to be sure a specific set of guarantees are provided by a 🟢 CAT. I think it's a bit out of the scope of this PR but it'd deserve an issue so that we can decide what a basic CAT runs.

erohmensing replied:
I'm more on your side for this question but I'm interested in hearing @evantahler's take on this specific question re: removing should_validate_schema

I agree that we should have a separate issue determining which tests should be canonically skippable

evantahler replied:
Which connector opts out of the `should_validate_schema` test today?  I can't find it ([search](https://github.com/search?q=repo%3Aairbytehq%2Fairbyte+should_validate_schema&type=code)).

+1 that we can revisit the choice to make some tests opt-out-able or not in another PR.   I'm comforted that at the very least, GA connectors need high-strictness, and won't opt out of any checks. 

brainstorming: I could imagine some source that doesn't have structured records (mongo?) that might have trouble with `should_validate_schema=True`... but then again, it would `fail_on_extra_fields` too.

erohmensing replied:
Search should be just for `validate_schema` which is the input key. [Here](https://github.com/airbytehq/airbyte/blob/6a10ae3e05fc4499fc2d6aa86fb0279050d36350/airbyte-integrations/connectors/source-azure-table/acceptance-test-config.yml#L18) is the connector

erohmensing replied:
Trying to run it to see if it would succeed without that, but its failing locally anyway 🙃

erohmensing replied:
Iiinteresting. But yeah it seems like it would fail the new test too, hmm 

It seems like in the case that we say don't do schema validation, we probably want to also skip failing on extra fields. But the other way around is not necessarily the case - if we skip failing on extra fields bc the connector hasn't been updated to fix it yet, we still want to run schema validation like we have been

erohmensing replied:
To me that means keeping 2 separate flags but letting the user know that the value of fail_on_extra_fields will always resolve to false if validate_schema is false

Potentially out of scope of this PR - seeing as we were questioning why a connector would want to bypass schema validation, there should def be bypass_reasons for declaring that

erohmensing replied:
In this case I'm leaving in the note about eventually sunsetting fail_on_extra_fields - I think we only expect that sources that need to skip schema validation would expect to fail on extra fields, and we always run it within schema validation.

erohmensing commented:
Updated the dockerfile and changelog (and docs) according to @alafanechere's suggestions. 

Added the bypass reasons for alpha and GA connectors that failed the test due to additional properties. Would appreciate a quick lookover @evantahler! Will also test a couple just to be sure. It's late so I'll come back and update the PR description tomorrow.

erohmensing commented:
Sneak peek into how I ran these things lives in this PR, which is still rough and hacky and will be cleaned up tomorrow: https://github.com/airbytehq/airbyte/pull/24377/files

The output I organized quasi-manually. More info [here](https://github.com/airbytehq/airbyte/issues/24316#issuecomment-1478171088) as well as a link to the results pushed to a remote commit.  

All in all, 91 connectors failed because of the test itself, of which the 26 that were updated here were GA or Beta. 
47 connectors actually ran the validation, output records, and those records succeeded. The rest failed for other various reasons (no acceptance-test-config.sh, no basic read test in the config, system_errors indicating something went wrong during `read` (these might rewquire further investigation)? ....

evantahler commented on docs/connector-development/testing-connectors/connector-acceptance-tests-reference.md:
👍 doc update

evantahler commented on airbyte-integrations/connectors/source-amazon-ads/acceptance-test-config.yml:
```diff
@@ -28,6 +28,7 @@ acceptance_tests:
       timeout_seconds: 2400
       expect_records:
         path: integration_tests/expected_records.jsonl
+      fail_on_extra_columns: false
```
lol, this was the only 'clean' change

erohmensing replied:
I've learned that this is the key to reviewing changes like these! 
<img width="243" alt="image" src="https://user-images.githubusercontent.com/22333512/227818422-d95bead9-09cd-4251-8e5d-3781efdfb67f.png">

evantahler commented on airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:
```diff
@@ -758,6 +758,11 @@ def should_validate_schema_fixture(self, inputs: BasicReadTestConfig, test_stric
 
+    @pytest.fixture(name="should_fail_on_extra_columns")
+    def should_fail_on_extra_columns_fixture(self, inputs: BasicReadTestConfig):
+        # TODO (Ella): enforce this param once all connectors are passing, sunset additionalProperties on high-level object
```
I thought we wanted to keep ensuring that additionalProperties was true so we could upgrade the version of a connector and not fail a yet-to-be-migrated config?

evantahler replied:
I agree that we shouldn't make a change about additionalProperties in this PR 😄.  I was reacting to your comment mostly.  I _think_ we can remove the check once we have config migrations, (1) I'm low-confident in that answer and (2) that's not any time soon.

erohmensing replied:
Oh yeah no I was saying that half of my comment shouldn't be there (I already removed it) 

I think we can enforce the parameter (all sources should fail on extra columns if they run schema validation), but not actually stop users from putting `additionalProperties: true` in the top level of the spec, for the reasons you mentioned. :) 

Beleive it or not I think we're on the same page haha

evantahler commented on airbyte-integrations/connectors/source-chartmogul/acceptance-test-config.yml:
```diff
@@ -4,26 +4,27 @@ connector_image: airbyte/source-chartmogul:dev
+        extra_fields: no
+        exact_order: no
+        extra_records: yes
+      fail_on_extra_columns: false
```
I spot-checked these, but I have low confidence that I didn't miss a tab or space 💀

erohmensing commented:
/test connector=connectors/source-airtable

erohmensing commented on airbyte-integrations/connectors/source-trello/acceptance-test-config.yml:
```diff
@@ -5,40 +5,39 @@ test_strictness_level: "high"
   discovery:
     tests:
-      - config_path: "secrets/config.json"
-        backward_compatibility_tests_config:
-          disable_for_version: 0.1.6
```
I removed this because it yelled about duplicate keys - I think we can only have one? Whenever I see people updating this to disable for a new version, they remove the old disable_for.

erohmensing commented:
I double-checked the other categories (failed for another reason) for beta and GA connectors and rebuilt/reran them. Most fell into the "success" or "failed and needs bypass" and the only one which didn't is failing on connector-health. So I am relatively confident we caught what we could!

Going to get this in and work on 
* Get the list of definite successes to Michael
* Creating the issues to fix these bypassed connector

erohmensing commented:
Tried to run airtable in the PR because of single-use token shenanigans, but realized it wasn't using the local version of CAT. Now running here: https://github.com/airbytehq/airbyte/actions/runs/4504802239

erohmensing commented:
/test connector=connectors/source-linkedin-ads connector-acceptance-test-version=dev

alafanechere commented:
/publish connector=bases/connector-acceptance-test auto-bump-version=false

erohmensing commented:
@alafanechere thank you! Forget that one a lot 🙈
